src/extensions/AssetMigrationUpgradeable.sol

```diff
@@ -0,0 +1,183 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.27;
+
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+import { IWETH } from "src/interfaces/IWETH.sol";
+import { ErrLengthMismatch, ErrEmptyArray } from "src/utils/CommonErrors.sol";
+
+abstract contract AssetMigrationUpgradeable {
+  using SafeERC20 for IERC20;
+
+  /// @dev Error when the storage location is null
+  error ErrInvalidStorageLocation();
+  /// @dev Error when the caller is not authorized
+  error ErrUnauthorizedCaller(address expected, address caller);
+
+  /// @dev The native token indicator address
+  address internal constant _NATIVE_TOKEN_INDICATOR = address(0);
+
+  /// @custom:storage-location erc7201:ronin.bridge.AssetMigration
+  struct AssetMigrationStorage {
+    IWETH _wnt;
+    // Migrator address
+    address _addr;
+    // Pending migrator address
+    address _pendingAddr;
+  }
+
+  /// @dev Emitted when a new migrator is set.
+  event NewMigrator(address indexed by, address indexed migrator);
+  /// @dev Emitted when the wrapped native token is set.
+  event WrappedNativeTokenSet(address indexed by, address indexed wnt);
+  /// @dev Emitted when the migrator transfer process is started.
+  event MigratorTransferStarted(address indexed migrator, address indexed newMigrator);
+
+  /**
+   * @dev Modifier to check if the caller is the migrator.
+   */
+  modifier onlyMigrator() {
+    _requireMigrator();
+    _;
+  }
+
+  /**
+   * @dev Returns the wrapped native token.
+   * - For `RoninGatewayV3`, it MUST return the WRON token.
+   * - For `MainchainGatewayV3`, it MUST return the WETH token.
+   */
+  function _getWrappedNativeToken() internal view returns (IWETH) {
+    return _getAssetMigration()._wnt;
+  }
+
+  /**
+   * @dev Migrates the given tokens to the specified addresses.
+   *
+   * When the token is the native (i.e RON or ETH), it will be wrapped (i.e WRON, WETH).
+   *
+   * Requirements:
+   * - The caller must be the migrator.
+   * - The length of the arrays must be the same.
+   * - The length of the arrays must not be zero.
+   */
+  function adminMigrate(address[] calldata tos, IERC20[] calldata tokens, uint256[] calldata amounts) external onlyMigrator {
```

Comment:
Should we transfer with the bridge's balance instead of passing amounts?

Reply:
we will need to exclude pending unclaimed withdrawal asset allocations